### PR TITLE
lorawan: adding NVM support for LoRaWAN and a reference SE implementation

### PR DESCRIPTION
**Pull request by GiulianoFranchetto**
For now, I have removed all the cryptographic calls from the implementation, as it was originally done using Nordic specific modules (mostly HUK and CryptoCell).

- [x] Need to abstract the crypto layer to not store the keys unencrytped on the flash
- [x] Various crypto must be managed, with great modularity. That will come from the fact that crypto API are very low-level stuff, and vendor dependent. I don't think you will have a Zephyr abstraction layer that will fit all use cases.
- [ ] Cryptos may come from west sub-project, and not zephyr tree

A more low-level implementation with key encryption using a daughter key can be found on the NCS repo: https://github.com/nrfconnect/sdk-zephyr/pull/678

Signed-off-by: Giuliano Franchetto <giuliano.franchetto@intellinium.com>

### CODE REVIEW

**Review by Mani-Sadhasivam — CHANGES_REQUESTED** [review]
Can you please split this change into multiple commits for NVM and SE?

File: subsys/lorawan/nvm/lorawan_nvm_settings.c [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+	LoRaMacNvmData_t *nvm = mibReq.Param.Contexts;
+
+	LOG_DBG("Crypto version: %"PRIu32", DevNonce: %d, JoinNonce: %"PRIu32,
+		mibReq.Param.Contexts->Crypto.LrWanVersion.Value,
+		mibReq.Param.Contexts->Crypto.DevNonce,
+		mibReq.Param.Contexts->Crypto.JoinNonce);
+
+	/* Wait for LoRa to be idle */
+	while (LoRaMacStop() != LORAMAC_STATUS_OK) {
```
Comment (JordanYates) [inline subsys/lorawan/nvm/lorawan_nvm_settings.c:58]: Why does the MAC layer need to be stopped and restarted here?

File: subsys/lorawan/lorawan.c [hunk 2]
```diff
@@ -576,7 +576,7 @@ int lorawan_start(void)
 	return 0;
 }
 
-static int lorawan_init(const struct device *dev)
+int lorawan_init(void)
```
Comment (JordanYates) [inline subsys/lorawan/lorawan.c:583]: What is the reasoning behind making applications call this function instead of making the priority user-configurable with Kconfig?

If you stick with this approach, the sample needs to be updated to actually call this.

File: subsys/lorawan/lorawan.c [hunk 3]
```diff
@@ -591,6 +591,7 @@ static int lorawan_init(const struct device *dev)
 	macCallbacks.NvmDataChange = NULL;
 	macCallbacks.MacProcessNotify = OnMacProcessNotify;
 
+
```
Comment (JordanYates) [inline subsys/lorawan/lorawan.c:602]: Nit: newline

**Review by JordanYates — CHANGES_REQUESTED** [review]

File: subsys/lorawan/nvm/Kconfig [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+# Copyright (c) 2021 Intellinium <giuliano.franchetto@intellinium.com>
+# SPDX-License-Identifier: Apache-2.0
+
+choice LORAWAN_NVM
+	bool "LoRaWAN NVM backend"
+	default LORAWAN_NVM_SETTINGS if SETTINGS
+	default LORAWAN_NVM_NONE
+
+config LORAWAN_NVM_NONE
```
Comment (Mani-Sadhasivam) [inline subsys/lorawan/nvm/Kconfig:11]: I think we can get rid of this option.

File: subsys/lorawan/nvm/Kconfig [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# LoRaWAN Non Volatile Memory configuration options
+
+# Copyright (c) 2021 Intellinium <giuliano.franchetto@intellinium.com>
```
Comment (Mani-Sadhasivam) [inline subsys/lorawan/nvm/Kconfig:3]: 2022 everywhere

File: subsys/lorawan/nvm/lorawan_nvm_settings.c [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+ * Copyright (c) 2021 Intellinium <giuliano.franchetto@intellinium.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <LoRaMac.h>
+#include <settings/settings.h>
+#include "lorawan_nvm.h"
+#include <logging/log.h>
+#if defined(CONFIG_LORAWAN_SE_HAS_OWN_NVM)
```
Comment (Mani-Sadhasivam) [inline subsys/lorawan/nvm/lorawan_nvm_settings.c:11]: Can you please make this commit as an independent PR? We should merge this one asap and backport to LTS.

File: subsys/lorawan/nvm/lorawan_nvm_settings.c [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+		mibReq.Param.Contexts->Crypto.LrWanVersion.Value,
+		mibReq.Param.Contexts->Crypto.DevNonce,
+		mibReq.Param.Contexts->Crypto.JoinNonce);
+
+	/* Wait for LoRa to be idle */
+	while (LoRaMacStop() != LORAMAC_STATUS_OK) {
+		k_sleep(K_MSEC(100));
+	}
+
+	SAVE_SETTINGS(nvm, LORAMAC_NVM_NOTIFY_FLAG_CRYPTO, Crypto)
```
Comment (Mani-Sadhasivam) [inline subsys/lorawan/nvm/lorawan_nvm_settings.c:62]: Don't we need locking here as restore can happen in parallel?

**Review by Mani-Sadhasivam — CHANGES_REQUESTED** [review]
@GiulianoFranchetto Sorry taking time to respond to this PR. You should consider splitting these two commits into separate PRs so that we can merge NVM support first and backport to LTS.

**Comment by JordanYates** [comment]
I should also note that the proposed API does not expose enough information for the application for it to decide when the Join EUI needs to be rotated (when DevNonce reaches UINT16_MAX)

File: subsys/lorawan/nvm/lorawan_nvm_settings.c (hunk 1 above)
Comment (GiulianoFranchetto) [reply subsys/lorawan/nvm/lorawan_nvm_settings.c:58]: This is taken from the loramac-node [sample app](https://github.com/zephyrproject-rtos/loramac-node/blob/zephyr/src/apps/LoRaMac/common/NvmDataMgmt.c#L72).
I guess it is to prevent more IRQ to be fired which would also change the NVM values at the same time, inducing race condition.
However, I'm not an expert in the LoRaMAC layer, so that could be unnecessary.

File: subsys/lorawan/nvm/lorawan_nvm_settings.c [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+K_WORK_DEFINE(nvm_settings_worker, nvm_settings_work);
+
+void lorawan_nvm_data_mgmt_event(uint16_t notifyFlags)
+{
+	LOG_WRN("New NVM notify flag: 0x%04X", notifyFlags);
+
+	nvm_notify_flag = notifyFlags;
+
+	if (notifyFlags != LORAMAC_NVM_NOTIFY_FLAG_NONE) {
+		k_work_submit(&nvm_settings_worker);
```
[the remark this answers is not in the document]
Comment (GiulianoFranchetto) [reply subsys/lorawan/nvm/lorawan_nvm_settings.c:91]: ok, done

File: subsys/lorawan/lorawan.c (hunk 2 above)
Comment (GiulianoFranchetto) [reply subsys/lorawan/lorawan.c:583]: Ok, corrected. That might be a change I did during testing that may not be necessary now.
I'll need to do the tests again just to be sure there is no issue with the initialization order of all the layers needed for the LoRaWAN stack.
The main issue I've found in the in the current implementation is that all the layers are deeply coupled, making the initialization a bit tricky

File: subsys/lorawan/nvm/Kconfig (hunk 4 above)
Comment (GiulianoFranchetto) [reply subsys/lorawan/nvm/Kconfig:11]: What would happen if a user choose to use LoRaWAN without ```CONFIG_SETTINGS``` enabled ?
If we want to make LoRaWAN to only work if at least settings is enabled, should we ```select SETTINGS``` when enabling LoRaWAN ?

File: subsys/lorawan/nvm/Kconfig (hunk 5 above)
Comment (GiulianoFranchetto) [reply subsys/lorawan/nvm/Kconfig:3]: Happy new year to all of you !

Comment (GiulianoFranchetto) [reply subsys/lorawan/nvm/Kconfig:3]: ps: done

File: subsys/lorawan/nvm/lorawan_nvm_settings.c (hunk 6 above)
Comment (GiulianoFranchetto) [reply subsys/lorawan/nvm/lorawan_nvm_settings.c:11]: working on it

File: subsys/lorawan/nvm/lorawan_nvm_settings.c (hunk 7 above)
Comment (GiulianoFranchetto) [reply subsys/lorawan/nvm/lorawan_nvm_settings.c:62]: ```lorawan_nvm_data_restore()``` is only called when ```lorawan_init()``` is called.
As for ```lorawan_nvm_data_mgmt_event()```, I think the race condition is prevented thanks to [that](https://github.com/zephyrproject-rtos/zephyr/pull/40193#discussion_r785417057).

**Comment by GiulianoFranchetto** [comment]
I agree that this is not exposed by the API. But is this really necessary ?
With the regulatory duty cycle that any LoRa endpoint must respect, I would guess that the DevNonce would reach UINT16_MAX after a very very very long time.

**Comment by GiulianoFranchetto** [comment]
A draft PR with only NVM related changes has been created: #41903

File: subsys/lorawan/nvm/Kconfig (hunk 4 above)
Comment (JordanYates) [reply subsys/lorawan/nvm/Kconfig:11]: Definitely not, I for example will need to write a custom backend for this as we don't use the settings API

**Comment by JordanYates** [comment]
> I agree that this is not exposed by the API. But is this really necessary ?
With the regulatory duty cycle that any LoRa endpoint must respect, I would guess that the DevNonce would reach UINT16_MAX after a very very very long time.

That depends, the recommended maximum backoff averages at 30 minutes, which would overflow after 4 years. If the device is rebooting regularly however it could reach that value earlier.

File: subsys/lorawan/nvm/Kconfig (hunk 4 above)
Comment (Mani-Sadhasivam) [reply subsys/lorawan/nvm/Kconfig:11]: > Definitely not, I for example will need to write a custom backend for this as we don't use the settings API

Okay, in that case it makes sense!
